src/Generator/Types/Std/Stdlib.cs

```diff
@@ -266,50 +264,45 @@ public override void CSharpMarshalToManaged(CSharpMarshalContext ctx)
                 textGenerator.WriteLine($"if ({ctx.ReturnVarName} == {nullPtr})");
                 textGenerator.WriteOpenBraceAndIndent();
                 textGenerator.WriteLine($"{ctx.Parameter.Name} = default({Type.Desugar()});");
-                textGenerator.WriteLine("return;");
+                textGenerator.WriteLine($"return{(ctx.Function.ReturnType.Type.IsPrimitiveType(PrimitiveType.Void) ? "" : " default")};");
                 textGenerator.UnindentAndWriteCloseBrace();
             }
 
-            string encoding = GetEncodingClass(ctx.Parameter);
+            string encoding = GetEncoding().Name;
             string type = GetTypeForCodePoint(encoding);
-            textGenerator.WriteLine($"var __retPtr = ({type}*) {returnVarName};");
-            textGenerator.WriteLine("int __length = 0;");
-            textGenerator.WriteLine($"while (*(__retPtr++) != 0) __length += sizeof({type});");
+            var retPtr = Generator.GeneratedIdentifier($"retPtr{ctx.ParameterIndex}");
+            var length = Generator.GeneratedIdentifier($"length{ctx.ParameterIndex}");
+            textGenerator.WriteLine($"var {retPtr} = ({type}*) {returnVarName};");
+            textGenerator.WriteLine($"int {length} = 0;");
+            textGenerator.WriteLine($"while (*({retPtr}++) != 0) {length} += sizeof({type});");
 
             ctx.Return.Write($@"global::System.Text.Encoding.{
-                encoding}.GetString((byte*) {returnVarName}, __length)");
+                encoding}.GetString((byte*) {returnVarName}, {length})");
         }
 
-        private string GetEncodingClass(Parameter parameter)
+        private (Encoding Encoding, string Name) GetEncoding()
```

Comment:
Why have you complicated this method? You can as well use `Context.Options.Encoding` at the caller's site.

Reply:
By the way, even though I've written it, it can be much better. You can delete the entire method and then simply use:
```C#
string encoding = Context.Options.Encoding.GetType().FullName;
```
and then drop the manually written name-space:
```C#
ctx.Return.Write($@"global::{encoding}.GetString((byte*) {returnVarName}, __length)");
```

Reply:
First I would like to say that I've tried to keep nearly the same logic as before even tho I don't completely understand the logic behind it.

 Using Encoding = Unicode with a 8 bit array will end up using `MarshalAs(UnmanagedType.LPWStr)` which will end up storing a unicode string in a 8 bit char array and I don't know if that's what people would expect to happen since we could also store it as UTF8. Maybe it should be allowed to specify more than one encoding per char type or width.

> Why have you complicated this method? You can as well use `Context.Options.Encoding` at the caller's site.

> By the way, even though I've written it, it can be much better. You can delete the entire method and then simply use:
> 
> ```cs
> string encoding = Context.Options.Encoding.GetType().FullName;
> ```

The method is checking for a supported encoding and it's also looking for the name of the static variable that holds an instance to a specific Encoding, not just the type.

Reply:
I understand, it's fine. I'll fix it myself after merging.